[SYCL-MLIR] Opaque pointer support Polygeist-to-LLVM

sommerlukas opened the pull request:
First step to add opaque pointer support for the SYCL-MLIR project: Make the lowering patterns in the Polygeist-to-LLVM lowering compatible with opaque pointers. 

Partly resolves #8616.

sommerlukas commented:
I did not want to migrate all our components to opaque pointers in a single step, as the resulting PR would be huge. Instead, this is the first step, affecting only the Polygeist to LLVM lowering. 

As some of the patterns are shared with other lowerings and other tests/components still rely on this lowering to produce typed pointers, I had to introduce a "versioning". The existing patterns now have a "Old" suffix, and the patterns with the unmodified name have been changed to support opaque pointers. The `polygeist-to-llvm` pass additionally supports a `use-opaque-pointers` option (defaulting to `false`), so it can continue to produce typed pointers when other parts rely on that (e.g. `cgeist`). 

The patterns (and functions) with the "Old" suffix as well as the option can be removed as soon as we complete the transition to opaque pointers. They are unmodified compared to `sycl-mlir`, so do not need to be reviewed. 

Having the versioning and option also allow us to continue to produce typed pointers, in case other tools out-of-scope of this project (e.g., SPIR-V translator) still need to complete the transition. 

The LIT tests for the Polygeist to LLVM lowering already verify the correctness of the opaque pointer version.

whitneywhtsang commented on polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:
```diff
@@ -103,8 +459,10 @@ struct BaseSubIndexOpLowering : public ConvertOpToLLVMPattern<SubIndexOp> {
+            .Case<LLVM::LLVMArrayType>(
                 [&](auto t) { currType = t.getElementType(); })
+            .Case<LLVM::LLVMPointerType>(
+                [&](auto t) { assert(false && "Pointer type not allowed"); })
```
Why not allowed to have struct of pointer?

sommerlukas replied:
It is allowed to have a struct of pointer (`struct<ptr>`) and this change doesn't inhibit that. 

What doesn't work anymore due to opaque pointers, is a struct behind a pointer nested inside a struct, i.e. `struct<ptr<struct>>`. 

Before opaque pointers, it was possible to access the `i64` element in this example: `struct<ptr<struct<i64, double>>`, because we could inspect the type of the struct pointed to. 

With opaque pointer, all we will see is `struct<ptr>`, as the element type of the pointer (`struct<i64, double>` in the example) isn't present with opaque pointers, so we need to stop inspection here. Hence the change here.

sommerlukas commented on polygeist/test/polygeist-opt/bareptrlowering.mlir:
```diff
@@ -496,28 +491,12 @@ func.func private @subindexop_memref_nested_struct(%arg0: memref<4x!llvm.struct<
-// CHECK-NEXT:      llvm.return %[[VAL_3]] : !llvm.ptr<f32>
-// CHECK-NEXT:    }
-
-func.func private @subindexop_memref_nested_struct_ptr(%arg0: memref<4x!llvm.struct<(ptr<struct<(f32)>>)>>) -> memref<?xf32> {
```
As explained [here](https://github.com/intel/llvm/pull/8767#discussion_r1148892668), we cannot have structs pointed to by pointers inside another struct, as in this test case. Therefore the test case was removed.

sommerlukas replied:
Sure, I've added that in the latest commit.

victor-eds commented on polygeist/test/polygeist-opt/sycl/cast.mlir:
```diff
@@ -1,12 +1,12 @@
+// CHECK:           llvm.return %[[VAL_1]] : !llvm.ptr
 // CHECK:         }
 
 func.func @test1(%arg0: memref<?x!sycl_range_1_>) -> memref<?x!sycl_array_1_> {
```
Please, can you update tests in this file chaning:
```
"sycl.cast"(%arg) : func-type
```
with:
```
sycl.cast %arg : orig-type to target-type
```
?

victor-eds reviewed (commented):
Can we have the pointee type in all check lines with `GEPOp` operations?

etiotto commented on polygeist/test/polygeist-opt/sycl/cast.mlir:
```diff
@@ -1,45 +1,45 @@
-// RUN: polygeist-opt --convert-polygeist-to-llvm --split-input-file %s | FileCheck %s
+// RUN: polygeist-opt --convert-polygeist-to-llvm='use-opaque-pointers=1' --split-input-file %s | FileCheck %s
```
We need, for the time being, to have both the original test cases that use the default (non-opaque pointers) as well as the new one. The reason is that, until we switch, we want to ensure that code changes do not regress existing tests. 
WDYT ?

sommerlukas replied:
Yes, good idea. I've added back the original tests without modifications for opaque pointers, with `-typed-pointer`-suffix to their name.

etiotto commented on polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:
```diff
@@ -54,6 +54,362 @@ namespace mlir {
 
+// FIXME: All the following patterns with an "Old" suffix to their name should
+// be removed once we drop typed pointer support.
+struct BaseSubIndexOpLoweringOld : public ConvertOpToLLVMPattern<SubIndexOp> {
```
I am thinking that we might be using non-opaque ptr for a while. Duplicating the code force us to dual maintain the code. My preference would be to keep one version of the code and use the new option introduced in this PR to change code snippets locally where we need to. WDYT?

sommerlukas replied:
My motivation to use this approach (for this PR) was that we have some changes that are not as localized as a simple `if`, for example [this change here](https://github.com/intel/llvm/pull/8767/commits/235f911407c2189034635cf579f19a02e3ac703e#diff-c49fbe1ad98fd477ae47be29989f0179b5efe97eddc9e9c3d4ca6ded24cc26d1L56-L74). 
I also think that with this approach, it will be easier and less error-prone to remove typed pointer support in the future, as it will be sufficient to remove the old patterns and the logic to add them to the pass. 

From the git history, it also seemed to me that this file is not as frequently subject to changes as other files in the project. Before my changes, it hadn't been modified in almost a month (apart from a fix for the merge from `sycl`) and mainly for refactoring purposed.

For the next part of the changes (SYCL to LLVM lowering), I've chosen an approach close to what you suggest, because the changes are more localized than in this PR. 

I'm also a bit more optimistic when it comes to switching the default to opaque pointers, based on comments like [this](https://github.com/intel/llvm/pull/8286):
> All tools are switched to opaque pointers already except those we use in compilation for offload targets.

etiotto replied:
If we are able to switch to using opaque pointers relatively quickly we can live with some code duplication for a few weeks. 
We just have to be careful and remember to dual maintain the duplicated code if we fix problems in that part of the code. 
I am OK with the PR overall.

sommerlukas replied:
Agreed, we should remember that for PR reviews affecting these patterns.

sommerlukas commented:
I assume CI will currently failed, as I have rebased on top of https://github.com/intel/llvm/pull/8774 and the branch is currently still missing the fix from https://github.com/intel/llvm/pull/8817. Will need to rebase again once #8817 has been merged.

sommerlukas commented:
> I assume CI will currently failed, as I have rebased on top of #8774 and the branch is currently still missing the fix from #8817. Will need to rebase again once #8817 has been merged.

Rebased now that #8817 has been merged.
